pylib_3.9.7/df/AntibodyStructurePrediction.py

```diff
@@ -0,0 +1,223 @@
+import base64
+import os
+import gzip
+from typing import Optional
+
+from Bio.SeqRecord import SeqRecord
+
+from df.bio_helper import column_to_sequences, string_to_sequence
+from df.data_transfer import ColumnData, TableData, DataFunctionRequest, DataFunctionResponse, DataFunction, DataType, \
+                             string_input_field, boolean_input_field, integer_input_field, input_field_to_column, \
+                             Notification, NotificationLevel
+
+from ruse.bio.antibody import align_antibody_sequences, ANTIBODY_NUMBERING_COLUMN_PROPERTY
+from ruse.bio.bio_data_table_helper import sequence_to_genbank_base64_str
+
+from ImmuneBuilder import ABodyBuilder2
+from ImmuneBuilder.ABodyBuilder2 import header as ABB2_HEADER
+from ImmuneBuilder.refine import refine
+from ImmuneBuilder.util import add_errors_as_bfactors
+
+class AntibodyStructurePrediction(DataFunction):
+    """
+    Predicts antibody structure from heavy and light chain sequences
+    """
+
+    def _antibody_numbering(self, ab_sequences: list[Optional[SeqRecord]], num_scheme: str,
+                            cdr_def: str) -> (list[Optional[str]], dict):
+
+        if not cdr_def:
```

Comment:
Maybe we should update these to Enum.  Though I know I'm guilty in AntibodyNumbering.py

Reply:
> Rather than strings ('imgt', 'kabat', etc.), create an ENUM in the antibody numbering code, and then use that here (and elsewhere)?

It's just a thought- string enums would be cleaner IMO.  Not necessary for the PR though.

Reply:
I agree - it would clean this up nicely.  I'll go ahead and make those changes in this PR.  This one is destined for 4.4.0.0, so it isn't in a huge rush, and this shouldn't take much time to do.

Reply:
Updated to Enum, but may be over-engineered.   Making the initial definitions into an Enum cascaded into various other places.  I think what I have now is reasonable, but I appreciate your feedback.

Reply:
I think the Enum stuff looks good.  I assume you checked the Antibody Numbering data functions work ok with the Enums?  I don't see any reason they wouldn't

Reply:
Yes, Antibody Numbering in both Python environments works.  I ran all the tests for GlysadePython as well and Pairwise Sequence Alignment gives me an error - needle returns a non-zero exit status.  I'm probably missing an environment variable for that one.